### ISSUE TITLE
Server gets stuck in `Error while rebuilding count index Conflicts with pending transaction. Please abort.` on startup

### ISSUE THREAD

**Issue by ospaarmann** [dgraph-io/dgraph#2229]
Latest version of Dgraph, as from the docker compose file from the docs. On Mac OS 10.13.3 with 16GB RAM. When trying to start the server with `docker-compose up` it first logs the error message `Echo error from server:7080. Err: rpc error: code = Unavailable desc = all SubConns are in TransientFailure` and then goes into an infinite loop of `2018/03/14 19:51:08 index.go:597: Error while rebuilding count index Conflicts with pending transaction. Please abort.`

```
Starting exdgraph_server_1 ...
Starting exdgraph_zero_1 ...
Starting exdgraph_ratel_1 ... done
Attaching to exdgraph_zero_1, exdgraph_server_1, exdgraph_ratel_1
zero_1    | Setting up grpc listener at: 0.0.0.0:5080
zero_1    | Setting up http listener at: 0.0.0.0:6080
zero_1    | 2018/03/14 19:50:59 node.go:246: Found hardstate: {Term:8 Vote:1 Commit:564 XXX_unrecognized:[]}
zero_1    | 2018/03/14 19:50:59 node.go:258: Group 0 found 564 entries
zero_1    | 2018/03/14 19:50:59 raft.go:405: Restarting node for dgraphzero
zero_1    | 2018/03/14 19:50:59 raft.go:567: INFO: 1 became follower at term 8
zero_1    | 2018/03/14 19:50:59 raft.go:315: INFO: newRaft 1 [peers: [], term: 8, commit: 564, applied: 0, lastindex: 564, lastterm: 8]
zero_1    | Running Dgraph zero...
zero_1    | 2018/03/14 19:50:59 node.go:127: Setting conf state to nodes:1
zero_1    | 2018/03/14 19:50:59 pool.go:168: Echo error from server:7080. Err: rpc error: code = Unavailable desc = all SubConns are in TransientFailure
zero_1    | 2018/03/14 19:50:59 pool.go:118: == CONNECT ==> Setting server:7080
zero_1    | 2018/03/14 19:50:59 pool.go:168: Echo error from server:7080. Err: rpc error: code = Unavailable desc = all SubConns are in TransientFailure
zero_1    | 2018/03/14 19:50:59 pool.go:168: Echo error from server:7080. Err: rpc error: code = Unavailable desc = all SubConns are in TransientFailure
zero_1    | 2018/03/14 19:50:59 pool.go:168: Echo error from server:7080. Err: rpc error: code = Unavailable desc = all SubConns are in TransientFailure
zero_1    | 2018/03/14 19:50:59 pool.go:168: Echo error from server:7080. Err: rpc error: code = Unavailable desc = all SubConns are in TransientFailure
zero_1    | 2018/03/14 19:50:59 pool.go:168: Echo error from server:7080. Err: rpc error: code = Unavailable desc = all SubConns are in TransientFailure
zero_1    | 2018/03/14 19:50:59 pool.go:168: Echo error from server:7080. Err: rpc error: code = Unavailable desc = all SubConns are in TransientFailure
zero_1    | 2018/03/14 19:50:59 pool.go:168: Echo error from server:7080. Err: rpc error: code = Unavailable desc = all SubConns are in TransientFailure
zero_1    | 2018/03/14 19:50:59 pool.go:168: Echo error from server:7080. Err: rpc error: code = Unavailable desc = all SubConns are in TransientFailure
zero_1    | 2018/03/14 19:50:59 pool.go:168: Echo error from server:7080. Err: rpc error: code = Unavailable desc = all SubConns are in TransientFailure
ratel_1   | 2018/03/14 19:51:00 Listening on port 8000...
zero_1    | 2018/03/14 19:50:59 pool.go:168: Echo error from server:7080. Err: rpc error: code = Unavailable desc = all SubConns are in TransientFailure
zero_1    | 2018/03/14 19:50:59 pool.go:168: Echo error from server:7080. Err: rpc error: code = Unavailable desc = all SubConns are in TransientFailure
zero_1    | 2018/03/14 19:50:59 pool.go:168: Echo error from server:7080. Err: rpc error: code = Unavailable desc = all SubConns are in TransientFailure
zero_1    | 2018/03/14 19:50:59 pool.go:168: Echo error from server:7080. Err: rpc error: code = Unavailable desc = all SubConns are in TransientFailure
zero_1    | 2018/03/14 19:51:01 raft.go:749: INFO: 1 is starting a new election at term 8
zero_1    | 2018/03/14 19:51:01 raft.go:580: INFO: 1 became candidate at term 9
zero_1    | 2018/03/14 19:51:01 raft.go:664: INFO: 1 received MsgVoteResp from 1 at term 9
zero_1    | 2018/03/14 19:51:01 raft.go:621: INFO: 1 became leader at term 9
zero_1    | 2018/03/14 19:51:01 node.go:301: INFO: raft.node: 1 elected leader 1 at term 9
server_1  | 2018/03/14 19:51:08 gRPC server started.  Listening on port 9080
server_1  | 2018/03/14 19:51:08 HTTP server started.  Listening on port 8080
server_1  | 2018/03/14 19:51:08 groups.go:86: Current Raft Id: 1
server_1  | 2018/03/14 19:51:08 worker.go:99: Worker listening at address: [::]:7080
server_1  | 2018/03/14 19:51:08 pool.go:118: == CONNECT ==> Setting zero:5080
zero_1    | 2018/03/14 19:51:08 zero.go:323: Got connection request: id:1 addr:"server:7080"
zero_1    | 2018/03/14 19:51:08 zero.go:420: Connected
server_1  | 2018/03/14 19:51:08 groups.go:109: Connected to group zero. Connection state: member:<id:1 addr:"server:7080" > state:<counter:567 groups:<key:1 value:<members:<key:1 value:<id:1 group_id:1 addr:"server:7080" leader:true last_update:1520981391 > > tablets:<key:"_predicate_" value:<group_id:1 predicate:"_predicate_" space:80145 > > tablets:<key:"director" value:<group_id:1 predicate:"director" space:20268 > > tablets:<key:"dummy" value:<group_id:1 predicate:"dummy" space:17 > > tablets:<key:"firend" value:<group_id:1 predicate:"firend" space:54 > > tablets:<key:"friend" value:<group_id:1 predicate:"friend" space:263 > > tablets:<key:"friends" value:<group_id:1 predicate:"friends" space:339 > > tablets:<key:"identifier" value:<group_id:1 predicate:"identifier" space:782 > > tablets:<key:"name" value:<group_id:1 predicate:"name" space:57646 > > tablets:<key:"release_date" value:<group_id:1 predicate:"release_date" space:32592 > > tablets:<key:"revenue" value:<group_id:1 predicate:"revenue" space:25848 > > tablets:<key:"running_time" value:<group_id:1 predicate:"running_time" space:32564 > > tablets:<key:"starring" value:<group_id:1 predicate:"starring" space:20454 > > > > zeros:<key:1 value:<id:1 addr:"zero:5080" leader:true > > maxLeaseId:10000 maxTxnTs:10000 maxRaftId:1 >
server_1  | 2018/03/14 19:51:08 draft.go:140: Node ID: 1 with GroupID: 1
server_1  | 2018/03/14 19:51:08 node.go:231: Found Snapshot, Metadata: {ConfState:{Nodes:[1] XXX_unrecognized:[]} Index:407 Term:6 XXX_unrecognized:[]}
server_1  | 2018/03/14 19:51:08 node.go:246: Found hardstate: {Term:6 Vote:1 Commit:1407 XXX_unrecognized:[]}
server_1  | 2018/03/14 19:51:08 node.go:258: Group 1 found 1000 entries
server_1  | 2018/03/14 19:51:08 draft.go:661: Restarting node for group: 1
server_1  | 2018/03/14 19:51:08 raft.go:567: INFO: 1 became follower at term 6
server_1  | 2018/03/14 19:51:08 raft.go:315: INFO: newRaft 1 [peers: [1], term: 6, commit: 1407, applied: 407, lastindex: 1407, lastterm: 6]
server_1  | 2018/03/14 19:51:08 groups.go:296: Asking if I can serve tablet for: id
server_1  | 2018/03/14 19:51:08 mutation.go:156: Done schema update predicate:"id" value_type:STRING directive:INDEX tokenizer:"exact"
server_1  | 2018/03/14 19:51:08 index.go:597: Error while rebuilding count index Conflicts with pending transaction. Please abort.
server_1  | 2018/03/14 19:51:08 index.go:597: Error while rebuilding count index Conflicts with pending transaction. Please abort.
server_1  | 2018/03/14 19:51:08 index.go:597: Error while rebuilding count index Conflicts with pending transaction. Please abort.
server_1  | 2018/03/14 19:51:08 index.go:597: Error while rebuilding count index Conflicts with pending transaction. Please abort.
server_1  | 2018/03/14 19:51:08 index.go:597: Error while rebuilding count index Conflicts with pending transaction. Please abort.
server_1  | 2018/03/14 19:51:08 index.go:597: Error while rebuilding count index Conflicts with pending transaction. Please abort.
```

Sometimes in between (every couple of thousand lines) this error message appears:

```
server_1  | 2018/03/14 19:51:08 index.go:597: Error while rebuilding count index Conflicts with pending transaction. Please abort.
server_1  | 2018/03/14 19:51:08 index.go:597: Error while rebuilding count index Conflicts with pending transaction. Please abort.
server_1  | 2018/03/14 19:51:08 mutation.go:156: Done schema update predicate:"name" value_type:STRING directive:INDEX tokenizer:"exact" tokenizer:"term" count:true
server_1  | 2018/03/14 19:51:08 groups.go:296: Asking if I can serve tablet for: age
server_1  | 2018/03/14 19:51:08 mutation.go:156: Done schema update predicate:"age" value_type:INT directive:INDEX tokenizer:"int"
server_1  | 2018/03/14 19:51:08 mutation.go:156: Done schema update predicate:"friend" value_type:UID count:true
server_1  | 2018/03/14 19:51:08 groups.go:296: Asking if I can serve tablet for: dob
server_1  | 2018/03/14 19:51:08 mutation.go:156: Done schema update predicate:"dob" value_type:DATETIME
server_1  | 2018/03/14 19:51:09 mutation.go:156: Done schema update predicate:"id" value_type:STRING directive:INDEX tokenizer:"exact"
server_1  | 2018/03/14 19:51:09 index.go:597: Error while rebuilding count index Conflicts with pending transaction. Please abort.
server_1  | 2018/03/14 19:51:09 index.go:597: Error while rebuilding count index Conflicts with pending transaction. Please abort.
server_1  | 2018/03/14 19:51:09 index.go:597: Error while rebuilding count index Conflicts with pending transaction. Please abort.
server_1  | 2018/03/14 19:51:09 index.go:597: Error while rebuilding count index Conflicts with pending transaction. Please abort.
server_1  | 2018/03/14 19:51:09 index.go:597: Error while rebuilding count index Conflicts with pending transaction. Please abort.
server_1  | 2018/03/14 19:51:09 index.go:597: Error while rebuilding count index Conflicts with pending transaction. Please abort.
```

**Comment by pawanrawal** [comment]
> zero_1    | 2018/03/14 19:50:59 pool.go:168: Echo error from server:7080. Err: rpc error: code = Unavailable desc = all SubConns are in TransientFailure

These are temporary until leader election happens for Dgraph server. Nonetheless, we should do a better job at logging there.

> server_1  | 2018/03/14 19:51:09 index.go:597: Error while rebuilding count index Conflicts with pending transaction. Please abort.

Is the server stuck as in the queries or mutations don't progress? We should definitely log the predicate for which this error occurs.

**Comment by ospaarmann** [comment]
> Is the server stuck as in the queries or mutations don't progress? 

This happens after I restart it with `docker-compose up`.

**Comment by pawanrawal** [comment]
Ok, I think the server seems stuck till the RAFT entries are being replayed after the restart. The server should respond once that is done.

Also, what is happening here is that probably because of the line below and `txn.startTs < l.commitTs`, the transaction to update count indexes are getting aborted on restart. Instead of logging these errors, they should be returned to the user. I'll make a change regarding this.

https://github.com/dgraph-io/dgraph/blob/441a431350dddb4d91f759dd05967a02928eb866/posting/list.go#L327

**Comment by pawanrawal** [comment]
I have merged a fix in accd1bf8b74bd24eab941a90254b3a9728ad9018 after which error would be returned to the user instead of printing to console.
